### DISCUSSION TITLE
Is it possible to apply alias generator from pydantic to FastApi fields?

### DISCUSSION THREAD

**Discussion by mclate** [fastapi/fastapi#8780] [Questions]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import Body, Query
from pydantic import BaseConfig, BaseModel, Field


def to_camelcase(string: str) -> str:
    res = ''.join(word.capitalize() for word in string.split('_'))
    return res[0].lower() + res[1:]


class User(BaseModel):
    class Config(BaseConfig):
        alias_generator = to_camelcase
        allow_population_by_field_name = True

    first_name: str = Field()
    last_name: str = Field()
    

@app.post('/user')
def create_user(
    user: User = Body(...), 
    send_invite: bool = Query(...),
):
    ...
```

### Description

On backend we're using FastApi and native to python `snake_case` notation, however, on frontend - there is a JS with their `camelCase` notation. 
In the code above, client can submit body with either snake or camel case and pydantic will do the magic to understand it. However, for the query parameter it doesn't work this way: `send_invite` is expected to be in snake case only. So this would be a violation of the overall structure either on the UI (they'd have to use snake for query and camel for body) or in the api (snake for everything, but camel for the query).

Thus, i wonder, what would be the best way to work this around?

### Operating System

Other

### Operating System Details

Not applicable (doesn't matter)

### FastAPI Version

0.65.2

### Python Version

3.9

### Additional Context

Pydantic version: 1.8.2

**Comment by Mause** [comment]
This works, but is not quite as convenient:

```python
class CamelCaseModel(BaseModel):
    class Config(BaseConfig):
        alias_generator = to_camelcase
        allow_population_by_field_name = True


class User(CamelCaseModel):
    first_name: str = Field()
    last_name: str = Field()


class Model(CamelCaseModel):
    user: User = Body(...)
    send_invite: bool = Query(...)


@app.post("/user")
def create_user(model=Depends(Model)):
    return model
```

**Comment by havardthom** [comment]
I would love a FastAPI app wide `alias_generator` type functionality, to be applied to e.g. Query and Path parameters. Currently I am manually defining camelCase aliases for all Path and Query parameters in my app.

**Comment by dshepelev15** [comment]
I think you really need to try [fastapi_utils](https://github.com/dmontagu/fastapi-utils) with their example of [ApiModel](https://github.com/dmontagu/fastapi-utils/blob/master/fastapi_utils/api_model.py)

**Comment by mclate** [comment]
That's already being in use (not through `fastapi_utils` though). The issue is not with the models, but with the query arguments that are not part of such model and cannot benefit from `alias_generator`

**Comment by ghost** [comment]
@mclate
Any snippets or workarounds for alias_generator that we can use with Query or Field params?

**Comment by tc-imba** [comment]
> @mclate Any snippets or workarounds for alias_generator that we can use with Query or Field params?

I've studied on this for a while, there are several workarounds, but all not perfect. I'm going to write a pr to provide a perfect solution to it.

I'm going to share some of the workarounds here:

# polyfill `alias_generator`

## with a decorator

Since we also need to modify the generated schema, dependency is not useful here. I use a decorator  to update the `alias` attribute of `Query` and `Path`. Maybe some more types of Param can be updated as well.

I use the `makefun` module to modify the signature of the function. I use the `snake2camel` from the `fastapi_utils` module, you can use any function to generate the alias you prefer.

```python
from inspect import Parameter, signature
from makefun import wraps
from fastapi_utils.camelcase import snake2camel

def camelcase_parameters(func: Any) -> Any:
    func_sig = signature(func)
    parameters = list(func_sig.parameters.values())
    for i, parameter in enumerate(parameters):
        if (
            parameter.default
            and isinstance(parameter.default, (params.Query, params.Path))
            and parameter.default.alias is None
        ):
            parameter.default.alias = snake2camel(parameter.name, start_lower=True)

    new_sig = func_sig.replace(parameters=parameters)

    @wraps(func, new_sig=new_sig)
    def wrapper(*args: Any, **kwargs: Any) -> Any:
        if "camelcase_parameters_dependency" in kwargs:
            del kwargs["camelcase_parameters_dependency"]
        return func(*args, **kwargs)

    return wrapper
```

Add the `camelcase_parameters` decorator between the `@router.xxx` and the route function definition. Now all `Query` and `Path` without alias will have a generated one (the camalCase of the original name).

```python
@router.get("/test")
@camelcase_parameters
async def test_route(invitation_code: str = Query(...)):
    return {"invitation_code": invitation_code}
```

## monkey patch `fastapi` dependency

Another solution is to modify `fastapi` itself in `fastapi/dependencies/utils.py`. The functions in this file are mainly for parsing the dependencies. We can modify a function `get_param_field`.

On this line

https://github.com/tiangolo/fastapi/blob/58ab733f19846b4875c5b79bfb1f4d1cb7f4823f/fastapi/dependencies/utils.py#L381

We can find that `alias` is set from `param.name` if `field_info.alias` is not provided. If we change it into

```python
if field_info.alias:
    alias = field_info.alias
else:
    alias = snake2camel(param.name, start_lower=True)
```

Now a generated alias with snake2camel is set. 

<details>
<summary>The complete version of the patch is</summary>
<p>

```python
from fastapi.dependencies.utils import *
from fastapi.dependencies import utils

def get_param_field(
    *,
    param: inspect.Parameter,
    param_name: str,
    default_field_info: Type[params.Param] = params.Param,
    force_type: Optional[params.ParamTypes] = None,
    ignore_default: bool = False,
) -> ModelField:
    default_value = Required
    had_schema = False
    if not param.default == param.empty and ignore_default is False:
        default_value = param.default
    if isinstance(default_value, FieldInfo):
        had_schema = True
        field_info = default_value
        default_value = field_info.default
        if (
            isinstance(field_info, params.Param)
            and getattr(field_info, "in_", None) is None
        ):
            field_info.in_ = default_field_info.in_
        if force_type:
            field_info.in_ = force_type  # type: ignore
    else:
        field_info = default_field_info(default_value)
    required = default_value == Required
    annotation: Any = Any
    if not param.annotation == param.empty:
        annotation = param.annotation
    annotation = get_annotation_from_field_info(annotation, field_info, param_name)
    if not field_info.alias and getattr(field_info, "convert_underscores", None):
        alias = param.name.replace("_", "-")
    else:
        # THE PATCH IS HERE
        if field_info.alias:
            alias = field_info.alias
        else:
            alias = snake2camel(param.name, start_lower=True)
    field = create_response_field(
        name=param.name,
        type_=annotation,
        default=None if required else default_value,
        alias=alias,
        required=required,
        field_info=field_info,
    )
    field.required = required
    if not had_schema and not is_scalar_field(field=field):
        field.field_info = params.Body(field_info.default)

    return field

utils.get_param_field = get_param_field
```

</p>
</details>  


You should put these lines before importing your apis (or using any dependency). In my project, I put it the first line in the file with `app = FastAPI(...)`.

However, this method is quite dirty. You have to fix to a `fastapi` release because if the function is modifed in a future release, it may introduce bad behaviors.

**Comment by tc-imba** [comment]
By default, with the polyfill I mentioned before `fastapi` will **ONLY** use the alias as a query or path string. 

For example, if the param is `invitation_code`, you can only use `GET /test?invitationCode=xxx` now in the previous example.

It is because functionality of `allow_population_by_field_name` is not supported. I also discover some methods to support it (to support using `GET /test?invitation_code=xxx` at the same time as an alternative).

# polyfill `allow_population_by_field_name`

In my project, I only polyfill `Query`, but I think other params will be similar.

## dependency

We can modify the query params in a dependency. But we need to consider the dependency resolve order: it should be resolved first before other `Query` dependency are resolved.

```python
from starlette.datastructures import MultiDict

def camelcase_parameters_dependency(request: Request) -> None:
    query_params = request.query_params
    new_params = MultiDict()
    for k, v in query_params.multi_items():
        if "_" in k:
            camel = snake2camel(k, start_lower=True)
            new_params.append(camel, v)
        else:
            new_params.append(k, v)
    request._query_params = new_params
```

You can add it as a dependency in your route:

```python
@router.get("/test", dependencies=[Depends(camelcase_parameters_dependency)])
@camelcase_parameters
async def test_route(invitation_code: str = Query(...)):
    return {"invitation_code": invitation_code}

```

Or you can add it in the router level dependency, or app level dependency, for example,

```python
app = FastAPI(dependencies=[Depends(camelcase_parameters_dependency)])
```

## middleware

Adding a middleware is quite similar to adding an app level dependency. I'll provide a simple example. It works but not well-written, please only take it as a reference.

```python
async def query_snake2camel_middleware(
    request: Request, call_next: Any
) -> JSONResponse:
    query_string = request.scope["query_string"]
    if query_string:
        params = parse_qs(request.scope["query_string"], keep_blank_values=True)
        d = {}
        for k, v in params.items():
            snake = k.decode("ascii")
            if "_" in snake:
                camel = snake2camel(snake, start_lower=True)
                if camel not in params:
                    d[camel.encode("ascii")] = v
                    continue
            d[k] = v
        request.scope["query_string"] = urlencode(d, doseq=True).encode("ascii")
    return await call_next(request)

app.middleware("http")(query_snake2camel_middleware)
```

## add dependency in decorator

You may notice that if we add the dependency at route level to every router which use the decorator, it will be inconvenient. If we add the dependency globally (or use the middleware), the performance of all api endpoints may drop. One possible solution is to add the dependency dynamically in the decorator. 

<details>
<summary>Here is the full code of it:</summary>
<p>

```python
from inspect import Parameter, signature
from makefun import wraps
from fastapi_utils.camelcase import snake2camel

def camelcase_parameters(func: Any) -> Any:
    func_sig = signature(func)
    parameters = list(func_sig.parameters.values())
    start_index = -1
    for i, parameter in enumerate(parameters):
        if (
            parameter.default
            and isinstance(parameter.default, (params.Query, params.Path))
            and parameter.default.alias is None
        ):
            if start_index < 0:
                start_index = i
            parameter.default.alias = snake2camel(parameter.name, start_lower=True)

    if start_index >= 0:
        parameters.insert(
            start_index,
            Parameter(
                "camelcase_parameters_dependency",
                kind=Parameter.POSITIONAL_OR_KEYWORD,
                default=Depends(camelcase_parameters_dependency),
            ),
        )
    new_sig = func_sig.replace(parameters=parameters)

    @wraps(func, new_sig=new_sig)
    def wrapper(*args: Any, **kwargs: Any) -> Any:
        if "camelcase_parameters_dependency" in kwargs:
            del kwargs["camelcase_parameters_dependency"]
        return func(*args, **kwargs)

    return wrapper
```

</p>
</details>  


I haven't got it full tested, and I'm not sure whether using `POSITIONAL_OR_KEYWORD` here will introduce any other issues. If someone try to use it and find any issue, you can ask me for help.

## monkey patch `fastapi` dependency

Same as the previous monkey patch, we use the file `fastapi/dependencies/utils.py`.

https://github.com/tiangolo/fastapi/blob/58ab733f19846b4875c5b79bfb1f4d1cb7f4823f/fastapi/dependencies/utils.py#L584-L589

We can find that only `alias` is used to get data. We can add a fallback to use `name`.

```python
        if is_scalar_sequence_field(field) and isinstance(
            received_params, (QueryParams, Headers)
        ):
            value = received_params.getlist(field.alias)
            if not value:
                value = received_params.getlist(field.name)
            if not value:
                value = field.default
        else:
            value = received_params.get(field.alias)
            if value is None:
                value = received_params.get(field.name)
```

<details>
<summary>The full example:</summary>
<p>


```python
from fastapi.dependencies.utils import *
from fastapi.dependencies import utils

def request_params_to_args(
    required_params: Sequence[ModelField],
    received_params: Union[Mapping[str, Any], QueryParams, Headers],
) -> Tuple[Dict[str, Any], List[ErrorWrapper]]:
    values = {}
    errors = []
    for field in required_params:
        if is_scalar_sequence_field(field) and isinstance(
            received_params, (QueryParams, Headers)
        ):
            value = received_params.getlist(field.alias)
            if not value:
                value = received_params.getlist(field.name)
            if not value:
                value = field.default
        else:
            value = received_params.get(field.alias)
            if value is None:
                value = received_params.get(field.name)

        field_info = field.field_info
        assert isinstance(
            field_info, params.Param
        ), "Params must be subclasses of Param"
        if value is None:
            if field.required:
                errors.append(
                    ErrorWrapper(
                        MissingError(), loc=(field_info.in_.value, field.alias)
                    )
                )
            else:
                values[field.name] = deepcopy(field.default)
            continue
        v_, errors_ = field.validate(
            value, values, loc=(field_info.in_.value, field.alias)
        )
        if isinstance(errors_, ErrorWrapper):
            errors.append(errors_)
        elif isinstance(errors_, list):
            errors.extend(errors_)
        else:
            values[field.name] = v_
    return values, errors

utils.request_params_to_args = request_params_to_args

```

</p>
</details>  

You should put these lines before any import as well. It's another dirty method, take care if you wish to use it.

**Comment by mclate** [comment]
This is some great study on how to approach the issue! There is one thing that seem to be conceptually missing: all these examples replace existing variable with camelCased ones, making it impossible to use the original one. The great advantage of PyDantic here is that it would allow one to use both notations (making it friendly for python and js users)
I wonder if i'm asking for too much here?....

**Comment by tc-imba** [comment]
> This is some great study on how to approach the issue! There is one thing that seem to be conceptually missing: all these examples replace existing variable with camelCased ones, making it impossible to use the original one. The great advantage of PyDantic here is that it would allow one to use both notations (making it friendly for python and js users) I wonder if i'm asking for too much here?....

In my second comment about polyfill `allow_population_by_field_name`, it allows you to use both camelCase and snake_case at the same time, just like the `APIModel` in  `fastapi_utils`.

However, all of the approaches I listed are not eligant. I'm going to work out some better approaches if achievable.

**Comment by ghost** [comment]
Thank you so much @tc-imba, these seem like something I can use 😄

**Comment by MUCZ** [comment]
> ```python
>             and isinstance(parameter.default, (params.Query, params.Path))
> ```

https://github.com/tiangolo/fastapi/issues/3662#issuecomment-961837270
@tc-imba 
Excuse me, but what is the params here?
`            and isinstance(parameter.default, (params.Query, params.Path))`
